### PR TITLE
Parallax

### PR DESCRIPTION
**Pull request by Tayyyyyyy**
**What does this PR do:**
See #11342

~~The SQL changes will be added in later, once this is merge ready, since resolving SQL merge conflicts is kinda annoying.~~
SQL changes added.

image

Now with the Lavaland planet! Sprite by Keekonox, code by Floyd/Qustinnus

🆑 AffectedArc07, Keekenox, Floyd/Qustinnus
add: P A R A L L A X (That fancy space thing)
imageadd: Adds a new Parallax layer that resembles Lavaland (Lava Planet), it spawns on a random location near the station. You need your parallax on high to see it.
/🆑

### CODE REVIEW

**Comment by AffectedArc07** [comment]
Allowed

Also Conflicted (The PR that is)

**Comment by Kyep** [comment]
This was testmerged with no issues.

**Comment by Tayyyyyyy** [comment]
SQL changes added back in. There are a few ifs that have spaces after that I will get to, but otherwise this is review ready.

**Comment by AffectedArc07** [comment]
Good job on adding the planet part. I held off because LL wasn’t in, but now it makes sense to throw in.

**Comment by Tayyyyyyy** [comment]
Thanks! It was all Mark's idea 🙂

File: code/modules/client/client procs.dm [hunk 1]
```diff
@@ -427,6 +427,9 @@
 		GLOB.admins -= src
 	GLOB.directory -= ckey
 	GLOB.clients -= src
+	if(movingmob != null)
```
Comment (Couls) [inline code/modules/client/client procs.dm]: If(movingmob)

File: code/__HELPERS/icon_smoothing.dm [hunk 2]
```diff
@@ -156,9 +156,13 @@
 		var/list/U = list()
 		if(fixed_underlay)
 			if(fixed_underlay["space"])
-				U += image('icons/turf/space.dmi', SPACE_ICON_STATE, layer=TURF_LAYER)
+				var/image/I = image('icons/turf/space.dmi', SPACE_ICON_STATE, layer=TURF_LAYER)
```
Comment (Fox-McCloud) [inline code/__HELPERS/icon_smoothing.dm]: Might want to make this mutable appearance and cache it higher up; check out TG's implementation of this.

File: code/_onclick/hud/ghost.dm [hunk 3]
```diff
@@ -59,3 +59,9 @@
 	using = new /obj/screen/ghost/teleport()
 	using.screen_loc = ui_ghost_teleport
 	static_inventory += using
+
+/datum/hud/ghost/show_hud()
```
Comment (Fox-McCloud) [inline code/_onclick/hud/ghost.dm]: Hmmm, why is this needed when parent calls `update_parallax_pref`, which eventually uses `create_parallax`

File: code/_onclick/hud/hud.dm [hunk 4]
```diff
@@ -172,6 +172,7 @@
 	mymob.update_action_buttons(1)
 	reorganize_alerts()
 	reload_fullscreen()
+	create_parallax()
```
Comment (Fox-McCloud) [inline code/_onclick/hud/hud.dm]: not `update_parallax_pref`?

File: code/_onclick/hud/parallax.dm [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+	var/turf/T = get_turf(M)
+	if(is_station_level(T.z))
+		invisibility = 0
+	else
+		invisibility = INVISIBILITY_ABSTRACT
+
+/obj/screen/parallax_layer/planet/update_o()
+	return //Shit wont move
+
+/obj/screen/parallax_pmaster
```
Comment (Fox-McCloud) [inline code/_onclick/hud/parallax.dm]: Not sure the purpose of this one.

File: code/_onclick/hud/parallax.dm [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+		C.parallax_layers.len = C.parallax_layers_max
+
+	C.screen |= (C.parallax_layers + C.parallax_static_layers_tail)
+
+/datum/hud/proc/remove_parallax()
+	var/client/C = mymob.client
+	C.screen -= (C.parallax_layers_cached + C.parallax_static_layers_tail)
+	C.parallax_layers = null
+
+/datum/hud/proc/apply_parallax_pref()
```
Comment (Fox-McCloud) [inline code/_onclick/hud/parallax.dm]: Also been changed.

File: code/_onclick/hud/parallax.dm [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+		L.transform = newtransform
+		animate(L, transform = matrix(), time = T, easing = QUAD_EASING | (new_parallax_movedir ? EASE_IN : EASE_OUT), flags = ANIMATION_END_NOW)
+		if(new_parallax_movedir)
+			L.transform = newtransform
+			animate(transform = matrix(), time = T) //queue up another animate so lag doesn't create a shutter
+
+	C.parallax_movedir = new_parallax_movedir
+	if(C.parallax_animate_timer)
+		deltimer(C.parallax_animate_timer)
+	C.parallax_animate_timer = addtimer(src, .update_parallax_motionblur, min(shortesttimer, PARALLAX_LOOP_TIME), TIMER_CLIENT_TIME, C, animatedir, new_parallax_movedir, newtransform)
```
Comment (Fox-McCloud) [inline code/_onclick/hud/parallax.dm]: also updated.

File: code/_onclick/hud/parallax.dm [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+/obj/screen/parallax_layer/proc/update_o(view)
+	if(!view)
+		view = world.view
+	var/list/new_overlays = list()
+	var/count = Ceiling(view/(480/world.icon_size))+1
+	for(var/x in -count to count)
+		for(var/y in -count to count)
+			if(x == 0 && y == 0)
+				continue
+			var/image/I = image(icon, null, icon_state)
```
Comment (Fox-McCloud) [inline code/_onclick/hud/parallax.dm]: mutable appearance.

File: code/_onclick/hud/parallax.dm [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+	overlays = new_overlays
+	view_sized = view
+
+/obj/screen/parallax_layer/proc/update_status(mob/M)
+	return
+
+/obj/screen/parallax_layer/layer_1
+	icon_state = "layer1"
+	speed = 0.6
+	layer = 1255
```
Comment (Fox-McCloud) [inline code/_onclick/hud/parallax.dm]: uuuhhhh um... hmm.

**Comment by Fox-McCloud** [comment]
This seems to be an older implementation of parallax, that will work for the most part, but will have problems with observers. Either case, I'd recommend implementing the most recent version, which is mostly the same, but has a few differences.

This also doesn't tie this in with shuttles; if that could be done, great, but if it's too much of an issue, don't worry about it *too* much.

**Comment by Tayyyyyyy** [comment]
AA said something about needing dynamic shuttle areas first, so I haven't looked into it. Is it possible to do without dynamic shuttle areas?

Will update parallax to tg's latest version, though.

File: code/_onclick/hud/parallax.dm (hunk 5 above)
Comment (Tayyyyyyy) [reply code/_onclick/hud/parallax.dm]: It looks like @AffectedArc07 refactored some of the code in `create_parallax` into layers that get added in `parallax_static_layers_tail` instead (`parallax_pmaster` and `parallax_space_whitifier`. Not sure what the difference is between the two approaches

**Comment by Fox-McCloud** [comment]
Yeah, don't worry about the shuttles then; I knew there was *something* about that, but I forgot what it was; thanks for reminding me.

But yeah, once the other is addressed, that'll help a lot with the other problems TG originally had with parallax.

File: code/_onclick/hud/ghost.dm (hunk 3 above)
Comment (Tayyyyyyy) [reply code/_onclick/hud/ghost.dm]: Removed this and it continued to work fine.

**Comment by KasparoVy** [comment]
Mutable appearance ftw!

**Comment by Fox-McCloud** [comment]
ok, last thing, then I think this is ready.

`get_smooth_underlay_icon` is used by more than just a base level proc; it's overridden a few times; will have to look into that (and test that it doesn't break smoothing) before this can be merged.

Other than that, looks good to go!

**Comment by Tayyyyyyy** [comment]
Having issues with shuttle diagonal walls not smoothing properly in transit. Will try to tackle it again tomorrow.

**Comment by Tayyyyyyy** [comment]
Turns out the diagonal shuttle wall issues was because transit space had to be moved up to be above parallax, but that put it on the same layer as diagonal shuttle walls.

Easiest fix I could think of was to just implement shuttle parallax, and turns out it can work without dynamic shuttle areas!

bla

**Comment by KasparoVy** [comment]
Bloody awesome, nicely done
